Assets/CustomAssets/Textures/MovingLava.shader

```diff
@@ -19,94 +22,115 @@
 			}
 
-			float2 random2(float2 p)
+			float2 random2d(float2 rand) //takes a vector2 to give an output of random vector2 between 0 and 1
 			{
-				//return frac(sin(float2(dot(p,float2(117.12,341.7)),dot(p,float2(269.5,123.3))))*43458.5453);
-				return frac(sin(float2(dot(p, float2(127.1, 311.7)), dot(p, float2(269.5, 183.3))))*43758.5453);
+				return frac(sin(float2(dot(rand, float2(127.1, 311.7)), dot(rand, float2(269.5, 183.3))))*43758.5453);
 			}
 
 			fixed4 frag(v2f i) : SV_Target
 			{
-				//fixed4 col = fixed4(0,0,0,1);
-				fixed4 col = fixed4(0,0,0,1);
 
-				float2 uv = i.uv;
-				uv *= _CellAmount; //Scaling cells
-				float2 iuv = floor(uv); //int values, no floating point
-				float2 fuv = frac(uv); //only the fractional part
-				float minDist = 1.0;
+				fixed4 col = _SecondaryColor;
+
+			//get world space light and view direction position
+			float3 viewDir = normalize(_WorldSpaceCameraPos.xyz - i.worldPos);
+			float3 lightDir = normalize(_WorldSpaceLightPos0.xyz);
+
+			float3 worldNormal = normalize(i.worldNormal);
+
+			float3 reflectionVector = reflect(-lightDir, worldNormal);
+
 
-				for (int y = -1; y <= 1; y++)
+			//find specular dot product
+			float RdotV = max(0, dot(reflectionVector, viewDir));
+
+			fixed4 specColor = _LightColor0 * pow(RdotV, _Shininess);
+
+
+			//Voronoi pattern
+			float2 grid = i.uv;
+
+			grid *= _CellAmount; //Scaling cells
+
+			//get values for the grid
+			float2 iuv = floor(grid); //int values, no floating point
+			float2 fuv = frac(grid); //only the fractional part
+
+			float minDistToCell = 1;
+
+			for (int y = -1; y <= 1; y++)
+			{
+				for (int x = -1; x <= 1; x++)
 				{
-					for (int x = -1; x <= 1; x++)
-					{
-						// Position of neighbour on the grid
-						float2 neighbour = float2(float(x), float(y));
-						// Random position from current + neighbour place in the grid
-						float2 pointv = random2(iuv + neighbour);
-						// Move the point with time
-						pointv = 0.5 + 0.5*sin(_Time.z + 6.2236*pointv);//each point moves in a certain way
-																		// Vector between the pixel and the point
-						float2 diff = neighbour + pointv - fuv;
-						// Distance to the point
-						float dist = length(diff);
-						// Keep the closer distance
-						minDist = min(minDist, dist);
-					}
+					// Position of neighbour cell on the grid
+					float2 neighbour = float2(x, y);
+
+					// Random point in relation to current and neighbour position on grid
+					float2 randomPoint = random2d(iuv + neighbour);
+
+					// Move the point with time, get the actual position of the offset cell (adding neighbour)
+					float2 cellPosition = 0.5 + _Speed * sin(_Time.z + _Offset * randomPoint) + neighbour;
+
+					//Calculate distance from the point in relation to current cell to grid fractional point
+					float2 difference = cellPosition - fuv;
+					float dist = length(difference);
+
+					// Keep the closer distance
+					minDistToCell = min(minDistToCell, dist);
 				}
```

Comment:
I think it would be better to put it to a separate function (like the `random2d` before). It would reduce the size of the main function and would help for readability (big chunks in main, while small chunks in separate functions)

Reply:
that function could be void if nothing has to be returned